File: scripts/compute_spectrum.m
%function compute_spectrum()
alpha=[-1,-0.5,0,0.5];
geo="emboite";
ni = 3;
type = 1;

matrix_P="P_"+geo+"_"+ni+"_"+type;
matrix_MTF=strings(1,size(alpha,2)-1);
load("../output/matrices/P_"+geo+"_"+ni+"_"+type+".mat");
nrows = size(eval(matrix_P),1);
matrix = zeros(nrows,size(alpha,2)*2);
headers = [];
    
    for i=0:size(alpha,2)-1
        disp(ni+" "+geo+" "+type+" "+alpha(i+1))
        load("../output/matrices/MTF_"+geo+"_"+ni+"_"+type+"_"+i+".mat");
        matrix_MTF(i+1)="MTF_"+geo+"_"+ni+"_"+type+"_"+i;
        [V,D]=eig(eval(matrix_MTF(i+1)),eval(matrix_P));
                
        z_real = real(diag(D));
        z_imag = imag(diag(D));
        matrix(:,i*2+1)=z_real(:);
        matrix(:,(i+1)*2)=z_imag(:);
        headers = [headers [alpha(i+1)+"_Real",alpha(i+1)+"_Imag"]];
        
    end
    
headers = strjoin(headers, ',');

%write header to file
fid = fopen("../output/csv/spectrum_"+geo+"_"+ni+"_"+type+".csv",'w'); 
fprintf(fid,'%s\n',headers);
fclose(fid);

%write data to end of file
dlmwrite("../output/csv/spectrum_"+geo+"_"+ni+"_"+type+".csv",matrix,'-append');

File: scripts/compute_spectrum.py
#! /usr/bin/python3
import numpy as np
from scipy import linalg
import matplotlib.pyplot as plt
import argparse
import csv

#############################
#        Arguments          #
#############################

parser = argparse.ArgumentParser()
parser.add_argument("--ni",type=int)
parser.add_argument("--geo",choices=['emboite','non_emboite'],type=str)
parser.add_argument("--type",choices=[0,1],type=int)
args = parser.parse_args()
ni = args.ni
geo = args.geo
type = args.type

###########################
#  Chargement matrices    #
###########################
print("Chargement matrices...")
A = np.loadtxt('../output/matrices/A_'+geo+'_'+str(ni)+"_"+str(type)+'.txt').view(complex)
nc = int(np.sqrt(A.shape[0]))
print(nc)
A = np.reshape(A,(nc,nc))

####
M = np.loadtxt('../output/matrices/M_'+geo+'_'+str(ni)+"_"+str(type)+'.txt')
nc = int(np.sqrt(M.shape[0]))
M = np.reshape(M,(nc,nc))

####
P = np.loadtxt('../output/matrices/P_'+geo+'_'+str(ni)+"_"+str(type)+'.txt')
nc = int(np.sqrt(P.shape[0]))
P = np.reshape(P,(nc,nc))

###########################
#  Calcul valeurs propres #
###########################
print("Calcul valeurs propres...\n")
alphas = [-1,-0.5,0,0.5]



with open("../output/csv/spectrum_"+geo+"_"+str(ni)+"_"+str(type)+".csv", 'w') as csv_file:
    writer = csv.writer(csv_file)

    # Columns head
    writer.writerow([str(alpha)+"_"+xy for alpha in alphas for xy in ["Real","Imag"]])
    
    data = []
    for alpha in alphas:
        print(alpha)
        MTF = -0.5*(A-alpha*M-(1-alpha)*P)

        val = linalg.eig(MTF,P,right=False)
        x = val.real
        y = val.imag

        data.append(x)
        data.append(y)


    writer.writerows([list(i) for i in zip(*data)])

